proposals/text/4084-agent-demo.md

```diff
@@ -0,0 +1,191 @@
+- Title: Agent tools
+- Decision driver: Vladimir Blagojevic (with Julian Risch)
+- Start Date: 2023-02-06
+- Proposal PR: https://github.com/deepset-ai/haystack/pull/4084
+
+
+# Summary
+
+Agent system consists of an extendable set of modules (aka experts/tools) and a "router" dispatching natural
+language input to a module that can best respond to the given input. These modules can be:
+
+• Neural, including the general-purpose LLMs as well as other smaller, specialized LMs.
+
+• Symbolic, for example a math calculator, a currency converter, or an API call to a search engine.
+
+
+Current Haystack pipeline design approach fits nicely with these extension modules/tools. They can be tested in
+isolation and subsequently integrated into the agent. The agent can register these tools, understand their objectives
+and route the user input to the appropriate tool. The tools can be either Haystack pipelines or components.
```

Comment:
So any class that inherits `BaseComponent`? and would this include a `DocumentStore` as well?

Reply:
@mayankjobanputra I can respond here because we already covered it in the Agent proposal and PR. 🙂 All that is required at the moment is a run method with a `query: str` parameter so that the Agent can invoke the tool and pass on a query. That corresponds to the BaseComponent's abstract run method, yes.
That's not the case for DocumentStore and also not for some of the StandardPipelines and it won't work for these. Even BaseRetriever does not strictly implement BaseComponent's abstract run method but we still made it work. I expect the pipeline refactoring to improve the issues with the subclasses not implementing the abstract methods correctly.